src/cache.jl

```diff
@@ -3,43 +3,67 @@
 using Core.Compiler: retrieve_code_info, CodeInfo, MethodInstance, SSAValue, SlotNumber
 using Base: _methods_by_ftype
 
-const compilecache = Dict{UInt, Any}()
+using Serialization, Scratch
+
 const compilelock = ReentrantLock()
 
-@inline function check_cache(driver, spec, id; kwargs...)
-    # generate a key for indexing the compilation cache
-    key = hash(kwargs, id)
-    key = hash(spec.name, key)      # fields f and tt are already covered by the id
-    key = hash(spec.kernel, key)    # as `cached_compilation` specializes on them.
-    for nf in 1:nfields(spec.f)
-        # mix in the values of any captured variable
-        key = hash(getfield(spec.f, nf), key)
-    end
+# whether to cache compiled kernels on disk or not
+const disk_cache = Ref(false)
+
+@inline function check_cache(cache, @nospecialize(compiler), @nospecialize(linker),
+                             spec, prekey; kwargs...)
+    key = hash((spec, kwargs), prekey)
+    force_compilation = compile_hook[] !== nothing
 
     # NOTE: no use of lock(::Function)/@lock/get! to keep stack traces clean
     lock(compilelock)
     try
-        entry = get(compilecache, key, nothing)
-        if entry === nothing
-            entry = driver(spec; kwargs...)
-            compilecache[key] = entry
+        obj = get(cache, key, nothing)
+        if obj === nothing || force_compilation
+            asm = nothing
+
+            # can we load from the disk cache?
+            if disk_cache[] && !force_compilation
+                path = joinpath(@get_scratch!("kernels"), "$key.jls")
+                if isfile(path)
+                    try
+                        asm = deserialize(path)
+                        @debug "Loading compiled kernel for $spec from $path"
+                    catch ex
+                        @warn "Failed to load compiled kernel at $path" exception=(ex, catch_backtrace())
+                    end
+                end
+            end
+
+            # compile
+            if asm === nothing
+                asm = compiler(spec; kwargs...)
+                if disk_cache[] && !isfile(path)
+                    serialize(path, asm)
+                end
+            end
+
+            # link (but not if we got here because of forced compilation)
+            if obj === nothing
+                obj = linker(spec, asm)
```

Comment:
We also need the kwargs passed to the linker step since we do device-specific global handling.

Reply:
Hmm. I don't really like all that `kwarg` shuffling, especially here since you'd then be passing the same set to multiple functions.

Reply:
Ideally we'd pass the CompilerJob around, but that's probably too costly to construct for every kernel call.

Reply:
I suppose I could just pass a closure or a callable object instead.